Add warning for missing system fonts during PDF rendering

scyyh11 opened the pull request:
**Important:** External contributors are required to comply with the template.

# Add warning for missing system fonts during PDF rendering

- Introduced a new method `_check_missing_fonts` to identify non-embedded fonts and log warnings if found.
- Added `is_embedded` method to `PdfFont` class to determine font embedding status.
- Enhanced logging to include text samples using non-embedded fonts for better context.

## Description

This PR adds a warning system to alert users when PDF pages contain non-embedded fonts that may not render correctly if the required system fonts are not available. When `PdfPage.render()` is called, the implementation now:

1. Scans the page for text objects using non-embedded fonts
2. Extracts sample text content that uses each problematic font
3. Logs a warning message listing the fonts and the text content that may be missing from the rendered image

The warning helps users identify when text might be silently omitted from rendered images due to missing system fonts, which was previously undetectable.

**Implementation details:**
- Uses PDFium's `FPDFFont_GetIsEmbedded()` API to check font embedding status
- Creates a temporary textpage to extract text content from text objects
- Limits text samples to 3 per font (with "and X more" indicator) to keep warnings readable
- Truncates very long text samples (>100 chars) for readability
- Gracefully handles edge cases (empty text, extraction failures, missing API)

**Example warning output:**
```
Page contains non-embedded fonts that may not render correctly if not available on the system: Arial (used in: "Hello", "World", "Test"); TimesNewRoman (used in: "Document", "Title" (and 2 more)). Text using these fonts may be missing from the rendered image.
```

## Checklist

<!--
- Use the Preview tab to see how the PR will render.
- Answer the questions below, choosing the section(s) relevant to your PR. Non-applying sections shall be set to `closed`.
- Place an `x` in the [ ] for yes, leave it empty for no. If a question is not applicable, remove the [ ], but keep the message in place.
-->

<details open><summary>Helpers</summary>

- [x] This PR changes helpers code.
- [ ] The change comes with sufficient test cases to confirm correct functionality.
- [ ] Any new test files are under a suitable license and have been registered in `REUSE.toml`.

</details>

<details closed><summary>Setup</summary>

- This PR changes setup code (`setupsrc/`, `setup.py` etc.).
- I have at least skimmed through [Installation -> With setup][1] and subsections.
- I have tested the change does not break internal callers.
- I believe the change is maintainable and does not cause unreasonable complexity or code fragmentation.
- The change does not shift a maintenance burden or upstream downstream tasks. *Keep handlers generic, avoid overly downstream-specific or (for us) effectively dead code passages.*
- I have assessed that the targeted use case cannot reasonably be satisfied by existing means, such as [Caller-provided data files][2], or the change forms a notable improvement over possible alternatives.
- I believe the targeted use case is supported by pypdfium2-team. *Note that we may not want to support esoteric or artificially restricted setup envs.*

</details>

<details closed><summary>Other</summary>

- This PR changes other things, namely: ... <!-- sum up change (keyword/topic) -->

</details>

[1]: https://github.com/pypdfium2-team/pypdfium2?tab=readme-ov-file#from-the-repository--with-setup
[2]: https://github.com/pypdfium2-team/pypdfium2?tab=readme-ov-file#with-caller-provided-data-files

*mara004 edit: improve markdown formatting*

scyyh11 committed:
Add font embedding check: Warn on non-embedded fonts in PdfPage rendering

- Introduced a new method `_check_missing_fonts` to identify non-embedded fonts and log warnings if found.
- Added `is_embedded` method to `PdfFont` class to determine font embedding status.
- Enhanced logging to include text samples using non-embedded fonts for better context.

mara004 commented on src/pypdfium2/_helpers/page.py:
```diff
@@ -459,6 +459,32 @@ def render(
+            logger.warning(
+                f"Page contains non-embedded fonts that may not render correctly if not available on the system: {font_list}. "
+                f"Text using these fonts may be missing from the rendered image."
+            )
```
Thank you for providing this code, I was able to take a quick look after all.

Could you factor out this implementation into a separate method to avoid lengthening the main `page.render()` method? Also, it may be better to make this optional. Opt-in, I think, as this potentially involves many API calls.

mara004 replied:
Or maybe just have it as a standalone public API, which the caller can invoke before `.render()` ?

mara004 commented on src/pypdfium2/_helpers/page.py:
```diff
@@ -495,6 +521,64 @@ def _auto_bitmap_format(page, fill_color, grayscale, prefer_bgrx, maybe_alpha):
+                    try:
+                        # Create a PdfTextObj with textpage to extract text
+                        # We create a new wrapper with textpage since the original obj doesn't have one
+                        text_obj_with_textpage = PdfTextObj(obj.raw, textpage=textpage)
```
WDYT, should we maybe add a `textpage` passthrough option to `PdfPage.get_objects()` to avoid the need to re-instantiate the `PdfTextObj`s ?

mara004 commented on src/pypdfium2/_helpers/pageobjects.py:
```diff
@@ -251,6 +251,16 @@ def get_weight(self):
+            bool: True if the font is embedded in the PDF, False if it needs to be loaded from the system.
+        """
+        if not hasattr(pdfium_c, 'FPDFFont_GetIsEmbedded'):
+            # If the API is not available, assume embedded to avoid false warnings
```
Do you have a pointer in which pdfium version this API was added?

scyyh11 committed:
Enhance PdfPage font handling: Add methods for missing font checks and improve object retrieval

- Updated `get_objects` method to accept an optional `textpage` parameter for better text extraction performance.
- Introduced `get_missing_fonts` and `warn_about_missing_fonts` methods to identify and log non-embedded fonts on the page.
- Renamed `is_embedded` method in `PdfFont` class to `get_is_embedded` for consistency with naming conventions.

scyyh11 committed:
Integrate missing font warnings into PdfPage rendering process

- Added a call to `warn_about_missing_fonts` in the `PdfPage` class to check for non-embedded fonts during rendering.
- Updated the `_check_missing_fonts` method to use `get_is_embedded` for font embedding checks, ensuring consistency with recent changes.

scyyh11 commented on src/pypdfium2/_helpers/page.py:
```diff
@@ -444,6 +530,9 @@ def render(
         """
         
+        # Check for non-embedded fonts and warn if found
+        self.warn_about_missing_fonts()
```
@mara004 For now, I’m still calling this method before rendering. But if you feel it’s unnecessary or would prefer it to be opt-in only, I’m happy to change it.

scyyh11 committed:
Fold `_check_missing_fonts` into `get_missing_fonts`

scyyh11 commented on src/pypdfium2/_helpers/page.py:
```diff
@@ -336,6 +340,127 @@ def flatten(self, flag=pdfium_c.FLAT_NORMALDISPLAY):
+        
+        return missing_fonts
+    
+    
```
Should I also add it for the other file that I changed?

scyyh11 replied:
Since the change in that file is just a small helper function, I’ll skip that for now.

scyyh11 committed:
Add SPDX license identifiers and copyright notice to PdfPage class

scyyh11 committed:
Remove the warning for non-embedded fonts in the PdfPage class during rendering.

scyyh11 committed:
Declare the `max_samples` outside the loop.

mara004 commented on src/pypdfium2/_helpers/page.py:
```diff
@@ -336,6 +340,131 @@ def flatten(self, flag=pdfium_c.FLAT_NORMALDISPLAY):
+                except (PdfiumError, AttributeError):
+                    continue
+        except Exception:
+            pass
```
I might prefer to avoid broad exception catching here.
In particular, I'd say `Exception` and `AttributeError` are too generic and could theoretically mask e.g. a typo in the code.
Also, I think there is some redundancy in your current try/except hierarchy.

mara004 replied:
Even with `PdfiumError`, I'm not sure silently suppressing it is good.
Unless there's a concrete reason (like with `textpage.get_textobj()` and escape chars), it might fit the API design better not to do any exception catching here, and leave it up to the caller whether they want to guard for `PdfiumError`.

scyyh11 committed:
Move get_missing_fonts() and warn_about_missing_fonts() from PdfPage to PdfTextPage where they logically belong. Remove broad exception catching to let errors propagate naturally. Add comprehensive test coverage for the new methods.

scyyh11 commented:
> A small test case or something would be great, for coverage.
> 
> Also, this could be added to the `pypdfium2 render` CLI as an opt-out mechanism. (I can do that bit in a post-merge commit if you want.)

Sounds good, please add that to the render CLI in a follow-up commit.

mara004 commented:
Thank you very much, I think I'll be able to merge this tomorrow.

scyyh11 committed:
Remove unnecessary whitespace.

mara004 commented on src/pypdfium2/_helpers/page.py:
```diff
@@ -302,7 +305,7 @@ def get_objects(self, filter=None, max_depth=15, form=None, level=0):
             
             # Don't register as child object, because the lifetime of pageobjects that are part of a page is managed by pdfium. The parent page should remain alive while a pageobject is used, but it seems unjustified to store countless of weakrefs just to lock pageobjects when the parent page is closed.
-            helper_obj = PdfObject(raw_obj, page=self, pdf=self.pdf, container=form, level=level)
+            helper_obj = PdfObject(raw_obj, page=self, pdf=self.pdf, container=form, level=level, textpage=textpage)
```
I wonder what's the best solution here. Perhaps add a `textpage` argument to `PdfObject` that is ignored but can be picked up by subclass?
An alternative might be checking `helper_obj.type == pdfium_c.FPDF_PAGEOBJ_TEXT` and post-assigning the textpage in that case, but I'm -1 on pulling in type if-checks here.

mara004 commented on src/pypdfium2/_helpers/textpage.py:
```diff
@@ -283,6 +283,95 @@ def search(self, text, index=0, match_case=False, match_whole_word=False, consec
+                if text_content:
+                    missing_fonts[font_name].append(text_content)
+        
+        return missing_fonts
```
Taking a closer look at what this function is doing, this just collects all non-embedded fonts – but even if the font is installed, this would currently raise an (erroneous) warning, right?
Is there any chance of checking whether the font is installed or not?
I'm not experienced with that part of the API, but I seem to remember something along the lines of `fpdf_sysfontinfo.h`.

mara004 replied:
Also, I'm not convinced unlimited sample collection / extraction is good.
Maybe move the max_samples mechanism here and make it a parameter? Caller can set `math.inf` if they really want unlimited samples.

scyyh11 replied:
I'm afraid that we are not able to get the actual glyph or font file that the system used to render the text using the current API.

mara004 replied:
Oh, right, I didn't think about substitution...
Then maybe rename `get_missing_fonts()` to `get_nonembedded_fonts()` ?
Because a non-embedded font isn't necessarily misssing.

mara004 replied:
> I'm afraid that we are not able to get the actual glyph or font file that the system used to render the text using the current API.

Though, checking whether the non-embedded font is installed on the system might still be possible? Even if we can't find out about substitution, it would avoid erroneous warnings if the exact font is installed.

scyyh11 replied:
Sorry for the late response, I was having my finals last week. 

Yes, it is possible to check whether the non-embedded font is installed. However, from what I understand, pypdfium2 just delegates font lookup to PDFium, and PDFium in turn hands the request to the platform (fontconfig on Linux, GDI on Windows, CoreText on macOS, etc.) and uses whatever comes back. To check system fonts ourselves we’d have to:

- call into platform-specific APIs or tools (fontconfig on Linux, different APIs on Windows/macOS)

- somehow normalize the PDF font name to what those APIs expose (family vs full name vs PostScript name, localized names, style variants, etc.).

Even then, we still wouldn’t be guaranteed to be checking the same thing pypdfium2 uses internally, so the result could still disagree with what actually happens at render time. I’m a bit worried that this would add a fair amount of platform-specific complexity.

mara004 replied:
I was hoping pdfium might be exposing APIs for that job, but if it doesn't, I agree it's not feasible.

Lei suggested `FPDF_SYSFONTINFO` in https://crbug.com/464315000#comment4 but I'm not quite sure how it works or what exactly it does.

scyyh11 committed:
Renamed `get_is_embedded` to `is_embedded` in the `PdfFont` class.

mara004 commented:
Again I am not 100% sure if you used AI here or not, but since your other two PRs had AI involvement I will mark this as suspected AI too.

To state it again for any readers, AI is now banned from this project. See https://github.com/pypdfium2-team/pypdfium2/pull/413#issuecomment-4759951913
